### PR TITLE
Full Site Editing: Add basic feature flags support

### LINKED ISSUE #33240: Full Site Editing: Add feature flag system [5]
To support plugin development with features that may have different release styles, let's add a basic feature flag system. We'll need to support flags on the server, client.

Please only implement what we need, vs over-engineering for future cases. If we end up with needing configuration files, please stick with a readable flat config list!

See also: https://martinfowler.com/articles/feature-toggles.html

### PR DESCRIPTION
**Pull request by Copons**
#### Changes proposed in this Pull Request

* Add a very basic query param based feature flags system working in both PHP and JS.

This implementation uses the `?fse_flags` query param that can contain a comma-separated list of flags, or nothing to clear the saved flags.
It supports both enabling (by default) and disabling flags (by prepending `-` to the flag name), like in Calypso. E.g.
```
?fse_flags=enabledFlag,-disabledFlag
```

Flags are saved into a cookie that expires when closing the browser, or when explicitly sending an empty `?fse_flags`.

The flags are then passed to JS via `wp_localize_script`.

To check them:

```php
// PHP
require_once 'path/to/lib/feature-flags/feature-flags.php';

$feature_flags = A8C_Full_Site_Editing_Feature_Flags::get_instance();

if ( $feature_flags->is_enabled( 'flag' ) ) { /* ... */ }
```

```es6
// JS
import isEnabled from 'path/to/lib/feature-flags';

if ( isEnabled( 'flag' ) ) { /* ... */ }
```

#### Notes

I've willingly avoided config files because I'm not sure how to add support for environment-based flags.
As I'm going on leave soon, feel free to commandeer or kill this PR as y'all see fit. 🙂 

#### Testing instructions

There are no feature flags implemented yet, but this is a nice way to test the PR nonetheless.

* Follow the [FSE readme](https://github.com/Automattic/wp-calypso/blob/e759e056360b72576e124989c4174afef78bc1db/apps/full-site-editing/README.md#local-development), and setup the plugin.
* Open the block editor and observe `fullSiteEditing.featureFlags` in the browser console (it should be an empty array).
* Add `?fse_flags=flag1,-flag2` to the URL and reload.
* Now `fullSiteEditing.featureFlags` should be `{ flag1: true, flag2: false }`.
* Change the query params and make sure the `fullSiteEditing.featureFlags` keeps updating as expected.
* Finally, add `?fse_flags` (empty).
* Make sure `fullSiteEditing.featureFlags` is an empty array again.

Fixes #33240

### CODE REVIEW

File: apps/full-site-editing/full-site-editing-plugin/lib/feature-flags/feature-flags.php [hunk 1]
```diff
@@ -0,0 +1,48 @@
+<?php
+
+class A8C_Full_Site_Editing_Feature_Flags {
+	private $flags = array();
+
+	public function __construct() {
+		$this->set_flags();
```
Comment (rodrigoi) [inline apps/full-site-editing/full-site-editing-plugin/lib/feature-flags/feature-flags.php]: Should this be a singleton? So we don't set the flags multiple times during the request lifetime.

**Comment by rodrigoi** [comment]
This looks like a reliable approach and works as expected. What I'm not entirely sure of (because if my limited PHP knowledge) is if it'll be safer to use a static class or a singleton, so we don't write headers to the response at the wrong time, causing an error.

File: apps/full-site-editing/full-site-editing-plugin/lib/feature-flags/feature-flags.php (hunk 1 above)
Comment (obenland) [reply apps/full-site-editing/full-site-editing-plugin/lib/feature-flags/feature-flags.php]: I'd like that. It would probably be beneficial to base it on the plugin architecture that #33250 introduces, and make it available to all FSE plugins

Comment (Copons) [reply apps/full-site-editing/full-site-editing-plugin/lib/feature-flags/feature-flags.php]: > Should this be a singleton?

Yup, it makes sense.

> base it on the plugin architecture that #33250 introduces, and make it available to all FSE plugins

Yup, it makes sense x2. 🙂
I worked on this while y'all were doing #33250, but I had already in mind to wait and rebase on that after it got merged.

File: apps/full-site-editing/full-site-editing-plugin/lib/feature-flags/feature-flags.php [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+	public function get_flags() {
+		return $this->flags;
+	}
+
+	public function is_enabled( $flag ) {
+		if ( ! isset( $flag ) ) {
+			return false;
+		}
+		return in_array( $flag, $this->flags );
```
Comment (mmtr) [inline apps/full-site-editing/full-site-editing-plugin/lib/feature-flags/feature-flags.php]: This checks if the given flag exists as a value in the given array, but we need to check if the flag is a key of the array, and return its value (since that's what determines if the flag is enabled or not).

```suggestion
		return array_key_exists( $flag, $this->flags ) ? $this->flags[ $flag ] : false;
```

File: apps/full-site-editing/full-site-editing-plugin/lib/feature-flags/feature-flags.php [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+		$this->set_flags();
+	}
+
+	/**
+	 * Creates instance.
+	 *
+	 * @return \A8C_Full_Site_Editing_Feature_Flags
+	 */
+	public static function get_instance() {
+		if ( null === self::$instance ) {
```
Comment (mattwiebe) [inline apps/full-site-editing/full-site-editing-plugin/lib/feature-flags/feature-flags.php]: PHP also has a `is_null` function, just FYI

File: apps/full-site-editing/full-site-editing-plugin/lib/feature-flags/feature-flags.php [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+				return;
+			}
+
+			setcookie( 'fse_flags', $_GET[ 'fse_flags' ] );
+			$flags = $_GET[ 'fse_flags' ];
+		} else if( isset( $_COOKIE[ 'fse_flags' ] ) ) {
+			$flags = $_COOKIE[ 'fse_flags' ];
+		}
+
+		$flags = explode( ',', $flags );
```
[the remark this answers is not in the document]
Comment (mattwiebe) [reply apps/full-site-editing/full-site-editing-plugin/lib/feature-flags/feature-flags.php]: Hmm I haven't been able to duplicate that behaviour. I've also never seen that extra-space-before-a-comma signature before, but it makes some sense for disambiguation in this context. Curious what PHPCS thinks, but I haven't been able to get it working on my system (yet).

File: apps/full-site-editing/full-site-editing-plugin/lib/feature-flags/feature-flags.php [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+				setcookie( 'fse_flags', '', time() - 3600 );
+				$this->flags = array();
+				return;
+			}
+
+			setcookie( 'fse_flags', $_GET[ 'fse_flags' ] );
+			$flags = $_GET[ 'fse_flags' ];
+		} else if( isset( $_COOKIE[ 'fse_flags' ] ) ) {
+			$flags = $_COOKIE[ 'fse_flags' ];
+		}
```
Comment (dmsnell) [inline apps/full-site-editing/full-site-editing-plugin/lib/feature-flags/feature-flags.php]: let's watch our nesting as it can quickly become hard to follow

I like to use helper bindings/variables to make the repeated logic less noisy

```php
$flags = null;

$has_flags_arg    = isset( $_GET['fse_flags'] );
$has_flags_cookie = isset( $_COOKIE['fse_flags'] );

if ( $has_flags_arg && empty( $_GET['fse_flags'] ) ) {
	setcookie( 'fse_flags', '', time() - 3600 );
	$this->flags = array();
	return;
}

if ( $has_flags_arg ) {
	setcookie( 'fse_flags', $_GET['fse_flags'] );
	$flags = $_GET['fse_flags'];
} else if ( $has_flags_cookie ) {
	$flags = $_COOKIE['fse_flags'];
}
```

if this all ends up a bit confusing we can consider moving it into a helper method
```php
// no side-effects here
private function get_requested_flags() {
	if ( isset( $_GET['fse_flags'] ) && ! empty( $_GET['fse_flags'] ) ) {
		return $_GET['fse_flags'];
	}

	if ( isset( $_COOKIE['fse_flags'] ) ) {
		return $_COOKIE['fse_flags'];
	}

	return '';
}

// all side-effects here
private function set_flags() {
	$flags = $this->get_requested_flags();

	// update cookie with new information - this may differ
	// from the previous flags from our cookie and may be
	// removing flags that were present
	$cookie_expiration = empty( $flags ) ? time() - 3600 : 0;
	setcookie( 'fse_flags', $flags, $cookie_expiration );

	if ( empty( $flags ) ) {
		$this->flags = array();
		return;
	}

	foreach ( explode( ',', $flags ) as $flag ) {
		$is_removing = '-' === $flag[0];
		$name = $is_removing ? substr( $flag, 1 ) : $flag;

		$this->flags[ $name ] = ! $is_removing;
	}
}
```

File: apps/full-site-editing/full-site-editing-plugin/lib/feature-flags/feature-flags.php [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+	 *
+	 * @param  string  $flag Feature flag.
+	 * @return boolean
+	 */
+	public function is_enabled( $flag ) {
+		if ( ! isset( $flag ) ) {
+			return false;
+		}
+		return array_key_exists( $flag, $this->flags ) ? $this->flags[ $flag ] : false;
+	}
```
Comment (dmsnell) [inline apps/full-site-editing/full-site-editing-plugin/lib/feature-flags/feature-flags.php]: `isset()` can take many args here and won't throw an error for accessing nested properties so we can reduce the logic to a single expression…

```php
public function is_enabled( $flagName ) {
	return isset( $flagName, $this->flags[ $flagName ] ) && $this->flags[ $flagName ];
}
```

File: apps/full-site-editing/full-site-editing-plugin/lib/feature-flags/feature-flags.php (hunk 1 above)
Comment (vindl) [reply apps/full-site-editing/full-site-editing-plugin/lib/feature-flags/feature-flags.php]: Just noting that this was converted to singleton in previous commits.

File: apps/full-site-editing/full-site-editing-plugin/lib/feature-flags/feature-flags.php (hunk 2 above)
Comment (vindl) [reply apps/full-site-editing/full-site-editing-plugin/lib/feature-flags/feature-flags.php]: Resolved in 011120a668d1d0887a6a013658f1c06b012a69d5 (in a different way).

File: apps/full-site-editing/full-site-editing-plugin/lib/feature-flags/feature-flags.php (hunk 3 above)
Comment (vindl) [reply apps/full-site-editing/full-site-editing-plugin/lib/feature-flags/feature-flags.php]: Updated in 011120a668d1d0887a6a013658f1c06b012a69d5.

File: apps/full-site-editing/full-site-editing-plugin/lib/feature-flags/feature-flags.php (hunk 5 above)
Comment (vindl) [reply apps/full-site-editing/full-site-editing-plugin/lib/feature-flags/feature-flags.php]: Refactored according to the first suggestion in 011120a668d1d0887a6a013658f1c06b012a69d5.

The second one is slightly different in terms of functionality - it's not handling the case of removing all of the flags when empty parameter value is passed correctly. We'd have to split that check out of `get_requested_flags`, so probably not worth it.

File: apps/full-site-editing/full-site-editing-plugin/lib/feature-flags/feature-flags.php (hunk 3 above)
Comment (obenland) [reply apps/full-site-editing/full-site-editing-plugin/lib/feature-flags/feature-flags.php]: 😞

File: apps/full-site-editing/full-site-editing-plugin/lib/feature-flags/feature-flags.php (hunk 6 above)
Comment (vindl) [reply apps/full-site-editing/full-site-editing-plugin/lib/feature-flags/feature-flags.php]: Updated to something similar in 011120a668d1d0887a6a013658f1c06b012a69d5. I just extracted the return statement into separate line to make it more readable.

File: apps/full-site-editing/full-site-editing-plugin/lib/feature-flags/feature-flags.php [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+		return $this->flags;
+	}
+
+	/**
+	 * Check if a feature flag is enabled.
+	 *
+	 * @param  string $flagName Feature flag.
+	 * @return boolean
+	 */
+	public function is_enabled( $flagName ) {
```
Comment (obenland) [inline apps/full-site-editing/full-site-editing-plugin/lib/feature-flags/feature-flags.php]: I must have done something wrong in `phpcs.xml`, camelCase variable names should trigger something.

File: apps/full-site-editing/full-site-editing-plugin/lib/feature-flags/feature-flags.php [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+	 * @var A8C_Full_Site_Editing_Feature_Flags
+	 */
+	private static $instance = null;
+
+	/**
+	 * Feature flags.
+	 *
+	 * @var array
+	 */
+	private $flags = array();
```
Comment (obenland) [inline apps/full-site-editing/full-site-editing-plugin/lib/feature-flags/feature-flags.php]: What do you think about using array shorthands throughout this file? We specified min PHP at 5.6.20 for the plugin in the repo

File: apps/full-site-editing/full-site-editing-plugin/lib/feature-flags/feature-flags.php (hunk 7 above)
Comment (vindl) [reply apps/full-site-editing/full-site-editing-plugin/lib/feature-flags/feature-flags.php]: Updated in ca80013dca76a3546a24ecaf596636b074e25de7.

File: apps/full-site-editing/full-site-editing-plugin/lib/feature-flags/feature-flags.php (hunk 8 above)
Comment (vindl) [reply apps/full-site-editing/full-site-editing-plugin/lib/feature-flags/feature-flags.php]: Sure, I like them more anyway. Updated in ca80013dca76a3546a24ecaf596636b074e25de7.
